railgun/test/integration/ingress_controllers_test.go

```diff
@@ -0,0 +1,91 @@
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	k8sgen "github.com/kong/kubernetes-testing-framework/pkg/generators/k8s"
+)
+
+var (
+	// ingressTimeout is the maximum amount of time that the tests should wait for an Ingress record to be provisioned and the backend accessible.
+	ingressTimeout = time.Minute * 5
+
+	// ingressTimeoutTick is the time to wait between Ingress resource timeout checks
+	ingressTimeoutTick = time.Second * 1
+)
+
+func TestMinimalIngress(t *testing.T) {
+	ctx := context.Background()
+
+	// deploy a minimal deployment to test Ingress routes to
+	container := k8sgen.NewContainer("nginx", "nginx", 80)
+	deployment := k8sgen.NewDeploymentForContainer(container)
+	_, err := cluster.Client().AppsV1().Deployments("default").Create(ctx, deployment, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	// expose the deployment via service
+	service := k8sgen.NewServiceForDeployment(deployment, corev1.ServiceTypeLoadBalancer)
+	_, err = cluster.Client().CoreV1().Services("default").Create(ctx, service, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	// route to the service via Kong Ingress
+	ingress := k8sgen.NewIngressForService("/nginx", map[string]string{
+		"kubernetes.io/ingress.class": "kong",
+		"konghq.com/strip-path":       "true",
+	}, service)
+	cluster.Client().NetworkingV1().Ingresses("default").Create(ctx, ingress, metav1.CreateOptions{})
+
+	// wait for the ingress backend to be routable
+	u := proxyURL()
+	assert.Eventually(t, func() bool {
+		resp, err := http.Get(fmt.Sprintf("%s/nginx", u.String()))
+		if err != nil {
+			t.Logf("WARNING: error while waiting for %s: %v", u.String(), err)
+			return false
+		}
+		defer resp.Body.Close()
+		return resp.StatusCode == http.StatusOK
+	}, ingressTimeout, ingressTimeoutTick)
+}
+
+func TestHTTPSRedirect(t *testing.T) {
+	ctx := context.Background()
+	opts := metav1.CreateOptions{}
+
+	container := k8sgen.NewContainer("alsonginx", "nginx", 80)
+	deployment := k8sgen.NewDeploymentForContainer(container)
+	_, err := cluster.Client().AppsV1().Deployments("default").Create(ctx, deployment, opts)
+	assert.NoError(t, err)
+
+	service := k8sgen.NewServiceForDeployment(deployment, corev1.ServiceTypeClusterIP)
+	_, err = cluster.Client().CoreV1().Services("default").Create(ctx, service, opts)
+	assert.NoError(t, err)
+
+	ingress := k8sgen.NewIngressForService("/example", map[string]string{
+		"kubernetes.io/ingress.class":           "kong",
+		"konghq.com/protocols":                  "https",
+		"konghq.com/https-redirect-status-code": "301",
+	}, service)
+	_, err = cluster.Client().NetworkingV1().Ingresses("default").Create(ctx, ingress, opts)
+	assert.NoError(t, err)
+
+	assert.Eventually(t, func() bool {
+		u := proxyURL()
+		resp, err := http.Get(fmt.Sprintf("%s/example", u.String()))
```

Comment:
> The reason it was failing is that the proxy on the cluster has HTTPS disabled (explicitly), so the packets could never make it to the proxy server itself.

Ah, okay, TIL that the Go stdlib HTTP client is not like curl and handles redirects by default, which isn't what I want here: we care about the initial response, not wherever the request would end up. This _should_ work:

```suggestion

	client := &http.Client{
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	assert.Eventually(t, func() bool {
		u := proxyURL()
		resp, err := client.Get(fmt.Sprintf("%s/example", u.String()))
```

I'm now running into some other issue where the Kong deployment is no longer created at all, and am not sure what's causing that. It's baked into the KIND image, correct?

Wasn't sure if that was maybe tied to the KTF changes that happened this PR started (go.mod was still pointing to a version from the 17th and there'd been further changes today). I did a replace (`replace github.com/Kong/kubernetes-testing-framework => github.com/shaneutt/kubernetes-testing-framework v0.0.0-20210318144855-8506060a256f`) and attempted another run, which did spawn the Kong deployment, but failed due to some other issue. Subsequent runs stopped spawning the Kong deployment again.